src/main/java/io/aiven/elasticsearch/gcs/storage/GcsBlobStoreRepository.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2020 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.elasticsearch.gcs.storage;
+
+import javax.crypto.SecretKey;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import io.aiven.elasticsearch.gcs.GcsSettingsProvider;
+import io.aiven.elasticsearch.storage.io.CryptoIOProvider;
+
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.common.base.Strings;
+import org.elasticsearch.cluster.metadata.RepositoryMetaData;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.blobstore.BlobPath;
+import org.elasticsearch.common.blobstore.BlobStore;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.unit.ByteSizeUnit;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.repositories.RepositoryException;
+import org.elasticsearch.repositories.blobstore.BlobStoreRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GcsBlobStoreRepository extends BlobStoreRepository {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GcsBlobStoreRepository.class);
+
+    public static final Setting<String> BASE_PATH =
+            Setting.simpleString(
+                    "base_path",
+                    Setting.Property.NodeScope,
+                    Setting.Property.Dynamic
+            );
+    public static final String TYPE = "gcs-encrypted";
+    static final Setting<ByteSizeValue> CHUNK_SIZE =
+            Setting.byteSizeSetting(
+                    "chunk_size",
+                    new ByteSizeValue(100, ByteSizeUnit.MB),
+                    new ByteSizeValue(1, ByteSizeUnit.BYTES),
+                    new ByteSizeValue(100, ByteSizeUnit.MB),
+                    Setting.Property.NodeScope
+            );
+
+    static final Setting<String> BUCKET_NAME =
+            Setting.simpleString(
+                    "bucket_name",
+                    Setting.Property.NodeScope,
+                    Setting.Property.Dynamic
+            );
+
+    private static final String ENCRYPTION_KEY_FILE = "key.enc";
```

Comment:
It's a file name suffix, right? Maybe, `.key.enc` then?

Reply:
no this is the filename

Reply:
Ah, right!